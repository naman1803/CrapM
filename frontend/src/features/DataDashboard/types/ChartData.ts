// used for SeverityBarChart
export interface SeverityData {
    facilityOwner: string;
    severityLevel: string;
    severityScore: number;
}

// used for SeverityOverTimeChart
export interface SeverityOverTimeData extends SeverityData {
    date: string;
}

// one instance of ranking
export interface FacilityExceedanceData {
    facilityOwner: string;
    noOfExceedances: number;
}
