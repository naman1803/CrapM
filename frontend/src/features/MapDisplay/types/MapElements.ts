import { LatLngExpression } from 'leaflet';

export enum MapElementType {
  Marker = 'marker',
  Circle = 'circle',
}

// Represent a point on the map
interface MapElementBase {
  id: number;
  dataType: string;
  isSelectable: boolean;
}

interface MarkerElement extends MapElementBase {
  elementType: MapElementType.Marker;
  dataPoint: GeoPoint;
}

interface CircleElement extends MapElementBase {
  elementType: MapElementType.Circle;
  dataPoint: GeoCircle;
}

export interface GeoPoint {
  coordinates: LatLngExpression;
}

interface GeoCircle extends GeoPoint {
  radius: number;
  colour: string;
}

export interface LakeGeoPoint extends GeoPoint {
  id: number;
  name: string;
  additional_notes?: string;
}

export interface SewageGeoPoint extends GeoCircle {
  id: number;
  facility_owner: string;
}

export type MapElement = MarkerElement | CircleElement;
