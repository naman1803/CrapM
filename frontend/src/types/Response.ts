//all response data related to general api requests
export interface ResponseData {
  status: number;
}
//format of data returned from the api
export type ApiData = object;

//format of api response
export interface ApiResponse<T extends ApiData> extends ResponseData {
  data: T;
}
