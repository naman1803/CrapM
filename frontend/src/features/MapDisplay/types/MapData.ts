import { LatLngExpression } from 'leaflet';
import { MapElement } from './MapElements';

//used for maps
export interface MapData {
  points: MapElement[];
  selectedPoint: MapElement | null;
  center: LatLngExpression;
  title: string;
  mapStyles?: React.CSSProperties;
}
