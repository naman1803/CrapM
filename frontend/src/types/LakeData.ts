import { ApiData } from './Response';

export interface LakeData extends ApiData {
  id: number;
  lakeName: string;
  latitude: number;
  longitude: number;
  additionalNotes?: string | null;
}
