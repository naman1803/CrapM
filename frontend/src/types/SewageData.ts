import { ApiData } from './Response';

export interface SewageData extends ApiData {
  id: number;
  facilityOwner: string;
  siteAddress: string;
  contaminant: string;
  typeOfExceedance: string;
  exceedanceStartDate: string;
  exceedanceEndDate: string;
  contaminantLimit: number;
  contaminantUnit: string;
  limitFrequency: string;
  noOfExceedances: number;
  quantityMaximum: number;
  quantityMinimum: number | null;
}

export interface SewageCoordData extends ApiData {
  id: number;
  facilityOwner: string;
  latitude: number;
  longitude: number;
  contaminants?: string;
  radius: number;
  colour: string;
}

export interface SewageSeverityData extends ApiData {
  id: number;
  facilityOwner: string;
  contaminants: Contaminants;
}

export interface Contaminants {
  [key: string]: {
    averageSeverityScore: number;
    overallSeverityLevel: string;
    averagePercentExceedance: number;
    records: ContaminantRecord[];
  };
}

interface ContaminantRecord {
  date: string;
  severityScore: number;
  severityLevel: string;
}

export interface ContaminantSeverityData extends ApiData {
  id: number;
  facilityOwner: string;
  contaminant: string;
  exceedanceStart: string;
  exceedanceEnd: string;
  severityScore: number;
  severityLevel: string;
}
