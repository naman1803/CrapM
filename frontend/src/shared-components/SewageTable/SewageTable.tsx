// SewageTable.tsx
import React from 'react';
import {
  Container,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from '@mui/material';
import { SewageData } from '@type/SewageData';

interface SewageTableProps {
  sewageData: SewageData[];
  isLoading: boolean;
}

export type { SewageTableProps };

const SewageTable: React.FC<SewageTableProps> = ({ sewageData, isLoading }) => {
  return (
    <Container>
      <Typography variant='h4' component='h1' gutterBottom>
        Industrial Sewage Data Overview
      </Typography>
      {isLoading ? (
        <Typography>Loading...</Typography>
      ) : (
        <TableContainer
          sx={{
            backgroundColor: 'rgba(0, 0, 0, 0.75)',
            border: '1px solid white',
            maxHeight: '83vh',
            overflowY: 'auto',
          }}
        >
          <Table>
            <TableHead>
              <TableRow>
                <TableCell>ID</TableCell>
                <TableCell>Facility Owner</TableCell>
                <TableCell>Site Address</TableCell>
                <TableCell>Contaminant</TableCell>
                <TableCell>Type of Exceedance</TableCell>
                <TableCell>Exceedance Start Date</TableCell>
                <TableCell>Exceedance End Date</TableCell>
                <TableCell>Contaminant Limit</TableCell>
                <TableCell>Contaminant Unit</TableCell>
                <TableCell>Limit Frequency</TableCell>
                <TableCell>No. of Exceedances</TableCell>
                <TableCell>Quantity Maximum</TableCell>
                <TableCell>Quantity Minimum</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {sewageData.map((data) => (
                <TableRow key={data.id}>
                  <TableCell>{data.id}</TableCell>
                  <TableCell>{data.facilityOwner}</TableCell>
                  <TableCell>{data.siteAddress}</TableCell>
                  <TableCell>{data.contaminant}</TableCell>
                  <TableCell>{data.typeOfExceedance}</TableCell>
                  <TableCell>{data.exceedanceStartDate}</TableCell>
                  <TableCell>{data.exceedanceEndDate}</TableCell>
                  <TableCell>{data.contaminantLimit}</TableCell>
                  <TableCell>{data.contaminantUnit}</TableCell>
                  <TableCell>{data.limitFrequency}</TableCell>
                  <TableCell>{data.noOfExceedances}</TableCell>
                  <TableCell>{data.quantityMaximum}</TableCell>
                  <TableCell>{data.quantityMinimum}</TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )}
    </Container>
  );
};

export default SewageTable;
