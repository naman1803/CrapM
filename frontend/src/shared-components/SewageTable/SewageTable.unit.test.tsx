import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import SewageTable, { SewageTableProps } from './SewageTable';
import { SewageData } from '@/types/SewageData';
const mockSewageData: SewageData[] = [
  {
    id: 1,
    facilityOwner: 'Facility 1',
    siteAddress: 'Address 1',
    contaminant: 'Contaminant 1',
    typeOfExceedance: 'Exceedance 1',
    exceedanceStartDate: '2024-01-01T00:00:00Z',
    exceedanceEndDate: '2024-01-02T00:00:00Z',
    contaminantLimit: 5,
    contaminantUnit: 'mg/L',
    limitFrequency: 'Daily',
    noOfExceedances: 2,
    quantityMaximum: 10,
    quantityMinimum: 1,
  },
];

describe('SewageTable', () => {
  it('should render the table', () => {
    const { asFragment } = renderSewageTable({
      sewageData: mockSewageData,
      isLoading: false,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render empty table when no data is available', () => {
    const { queryAllByText } = renderSewageTable({
      sewageData: [],
      isLoading: false,
    });
    expect(queryAllByText(/Facility/)).toHaveLength(1);
  });

  it('should display loading message when loading', () => {
    const { getByText } = renderSewageTable({
      sewageData: [],
      isLoading: true,
    });
    expect(getByText('Loading...')).toBeInTheDocument();
  });
});

const renderSewageTable = (props: SewageTableProps) =>
  render(<SewageTable {...props} />);
