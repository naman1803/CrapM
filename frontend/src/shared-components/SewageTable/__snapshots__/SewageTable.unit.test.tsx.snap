// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SewageTable should render the table 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
  >
    <h1
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
    >
      Industrial Sewage Data Overview
    </h1>
    <div
      class="MuiTableContainer-root css-kp6v00-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Facility Owner
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Site Address
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Contaminant
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Type of Exceedance
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Exceedance Start Date
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Exceedance End Date
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Contaminant Limit
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Contaminant Unit
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Limit Frequency
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              No. of Exceedances
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Quantity Maximum
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Quantity Minimum
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Facility 1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Address 1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Contaminant 1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Exceedance 1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              2024-01-01T00:00:00Z
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              2024-01-02T00:00:00Z
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              5
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              mg/L
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Daily
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              10
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              1
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
